modding-tutorials/harmony-basics.md

```diff
@@ -0,0 +1,181 @@
+		foreach(MethodBase method in Harmony.GetAllPatchedMethods()){
+			//See if the one you need has been modified
+			if(method.Name == "OpenBox")
+				//Skip patch because another mod changed something you use
+				return false;
+		}
+		return true;
+	}
+
+```
+
+---
+###### [HarmonyPrefix]
+This code takes place just before the original method happens and can be used to skip it all together. 
+
+This means you can replace a method entirely by having a Prefix do stuff then skip the original method by returning `false`
+
+If the original method needed to return some value the patched method can accept the `ref "Type" __result` arguement by reference. This will make it look like the original method is returning whatever `__result` is set to
+```csharp
+	[HarmonyPrefix]
+	static bool AlwaysReturnTrue(ref bool __result)
+	{
+		//Set return value to true
+		__result = true;
+		//Tell Harmony to not run the original method
+		return false;
+	}
+```
+
+You can also use this method to store values that can be accessed in the Postfix if you have one via `__state`. You can use `ref` or `out` to set the `__state` to whatever value or type you want. If you need to store more data then you need to make your own object to pass on to the Postfix
+
+```csharp
+	static void Prefix(out Stopwatch __state)
+    {
+        __state = new Stopwatch(); // Stopwatch is a custom timer class
+        __state.Start();
+    }
+
+    static void Postfix(Stopwatch __state)
+    {
+        __state.Stop();
+        FileLog.Log(__state.Elapsed.ToString());
+    }
+```
+
+
+
+---
+###### [HarmonyPostfix]
+This code takes place right after the original method. It can take in the `__result` from the original method and use it or even change it. 
+
+One benefit of Postfixs is that they always run. No matter where the original method escaped its code from it will go to the Postfix right after. 
+
+{% hint style="info" %} One quirk of Postfixs is that they can either change `__result` by ref or can return a value that would work the same way, but if they return a result it has to be the same Type as whatever the first arguement in the method is! {% endhint %}
+
+```csharp
+	[HarmonyPostfix]
+	static BuildError CheckForAlt(BuildError __result, BlockCreator __instance)
+	{...}
+```
+
+---
+###### [HarmonyTranspiler]
+Transpilers are a much more advanced way to get your own code to run. It is a bit out of the scope of a tutorial here so I will link to [this article](https://harmony.pardeike.net/articles/patching-transpiler.html#basic-transpiler-tutorial)
+
+To give a simplified overview: Instead of writing code to be executed, you are looking at all of the IL instructions that a chunk of code represents and doing some of your own instructions whenever you think the right time is. 
+
+---
+###### [HarmonyTargetMethod]
+I'll be honest I don't understand this one and works other than it is a different way to tell Harmony which method you are changing instead of using the class attribute
```

Comment:
`[HarmonyTargetMethod]` can be used to dynamically choose a target method. If you can't explain this one, leave it out of the docs.